Feature/Audit [Update/Simpler version]

kaareal opened the pull request:
**This is not finished, i am sure there is bugs in it.**

Before I continue I want to check that we are roughly agree on the api interface.

- I am no sure if the PATCH /1/products/:id solutions is what you want
- And not sure if I need to adjust the handling of the POST /1/products/search 

Also note that I picked the products endpoint as an example I don't plan to add the audit logic to any endpoint out of the box.

dominiek commented on services/api/src/routes/products.js:
```diff
@@ -3,6 +3,8 @@ const Joi = require('joi');
 const validate = require('../utils/middleware/validate');
 const { authenticate, fetchUser } = require('../utils/middleware/authenticate');
 const { searchValidation, getSearchQuery, search } = require('../utils/search');
+const Audit = require('../utils/audit');
```
I understand you're choosing the word `Audit` here because it's short. But an `AuditTrail` is semantically very different than an `Audit`. An "audit" is really the analysis of an audit trail in order to create a report. So `Audit.create` is semantically bad. We're either:

- Adding to an "audit trail" or
- Appending an "audit trail" or
- Creating an "audit trail entry"

kaareal replied:
What about AuditLog would that be acceptable?

andrewplummer replied:
+1 for `AuditEntry` for me... yes it's semi-non-standard pluralization but that shouldn't be a barrier

dominiek commented on services/api/src/routes/products.js:
```diff
@@ -26,13 +28,15 @@ router
   )
   .get('/:productId', async (ctx) => {
     const { product } = await ctx.state;
+    await Audit.read(ctx, 'view product', product.id);
```
We're not really reading the audit (report) here nor are we reading the audit trail. We're creating an audit trail entry, so more correct and clear would be:

- `AuditTrail.add('read')`
- `AuditTrail.append('read')`
- `AuditTrailEntry.create('read')`
- `createAuditTrailEntry('read')`
- `appendAuditTrail('read')`

kaareal replied:
I was drawing parallels to logger.error doesn't throw an error but logs an error. Similar i would `Audit[type](something)`

Its not clear to me if you want the just one method for each for the different types or you are not seeing that I got a method for each type (because the interface is different)?

andrewplummer replied:
I think there's a lot of assumptions baked into the needs here and CRUD style verbs is one. A fair assumption is that we always have an action but that might not fall into CRUD. We have more assumptions too like there will always be a context object and there will always be an ID which I also think may or may not be true.

Given this I think defining our interface less strictly is a good idea so for example:

```
AuditEntry.create('read', {
   ctx, // <-- a "ctx" could be something that AuditEntry knows how to "unpack"
   objectId: product.id,
   userId: ctx.state.authUser.id,
});
```

or even:

```
AuditEntry.create({
   action: 'read',
   objectId: product.id,
   userId: ctx.state.authUser.id,
   ctx,
});
```

Even if `action` is required I think I prefer a single options object to multiple args for readability/composability.

kaareal replied:
yeah i think i am gonna go that way, its gets around the issue disagreement around naming. 

And can be added later.

dominiek commented on services/api/src/routes/products.js:
```diff
@@ -56,6 +60,12 @@ router
+
+      // XXX What todo here?
+      // 1. No object should we track create a db record for each ObjectId return from the query
+      // 2. Change the db model
```
object ID should be optional. It's only relevant to CREATE, UPDATE and DELETE

dominiek commented on services/api/src/routes/products.js:
```diff
@@ -79,6 +91,7 @@ router
   .delete('/:productId', async (ctx) => {
     const product = ctx.state.product;
     await product.delete();
+    await Audit.delete(ctx, 'delete product', product.id);
```
I think it might be better to have the action name in the past tense - it makes it easier to use in the admin UI:

- X deleted product Y

kaareal replied:
Yep for sure, note that this is just an example how it could be hooked up very little effort went into adding the action text.
As i mentioned above before this PR is done i will remove it audit util from this endpoint.

dominiek commented on services/api/src/utils/audit.js:
```diff
@@ -0,0 +1,58 @@
+  }
+}
+
+module.exports = Audit;
```
Btw, most of our helper libraries don't export like this. Shouldn't we align that?

kaareal replied:
The class gets in the way, I can naturally just split it apart into functions and get the same.

cc @andrewplummer

andrewplummer replied:
I'm still a fan of removing it entirely and just having a model :)

dominiek commented on services/api/src/utils/audit.js:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  static getDiff(object, fields) {
+    const pathsModified = object.directModifiedPaths();
```
Is this a mongoose facility? Nice

dominiek commented on services/api/src/models/definitions/audit-log.json:
```diff
@@ -0,0 +1,40 @@
+{
```
I usually try to avoid model names that don't pluralize well. The name of this model is plural.

I recommend using a singular noun that can be pluralized:

- `audit-trail-entry` / `audit-trail-entries`
- `audit-log-entry` / `audit-log-entries`

kaareal replied:
isnt `logs` the plural form? log would be singular.

if you dont like log what about record? 

i prefer the plural from is just adding an `s` to the end + short

andrewplummer commented on services/api/src/models/definitions/audit-log.json:
```diff
@@ -0,0 +1,40 @@
+    },
+    "type":  {
+      "type": "String",
+      "enum": ["create", "update", "delete", "read"]
```
maybe it doesn't need to be an enum, just a freeform field called `action`

kaareal replied:
fyi I do have a free form field called action. 
The `type` field was to todo make it easier to query on e.g. find me all deletions or all updates done to a given object

kaareal replied:
@dominiek
 
I think it depends on the `Regenerate Invoice` what does. Does it update the invoice object or does it create a new invoice object. (then it would either be update or create)

Let's say we are starting a background job e.g. generate Invoice`s`, where there is no objectId. 

I guess we could have an `generic` or `action`. Or just reuse e.g. the `create`.
 
I am having a hard time seeing what value it brings. Like does having a generic / action make it easier to query?

kaareal replied:
Thinking about it more ... this is suppose to be http based, we got the request method and url ... i dont think we need type at all.

andrewplummer commented on services/api/src/models/definitions/audit-log.json:
```diff
@@ -0,0 +1,40 @@
+      "type": "String",
+      "enum": ["create", "update", "delete", "read"]
+    },
+    "diffObject": {
```
Is just `diff` not enough? To me the object part is implied and other words are just confusing... I know what a diff is.

kaareal commented on services/api/src/models/definitions/audit-log.json:
```diff
@@ -0,0 +1,40 @@
+    "diffObject": {
+      "type": "Mixed"
+    },
+    "objectId": {
```
is there a reason you don't want to aligned it with naming from mongo?

kaareal commented:
I have updated the PR, renamed a few things.

The modal fields has changed a bit, besides from changing the interface + moving it to be just methods on db model.

## Add routePrefix (namespace)
This is gonna require a wide change e.g. that relies on the router.opts.prefix see https://github.com/koajs/router/blob/master/API.md#router- from what i can tell we doing our routes wrong ... but that is a separate discussion. This field is only set if you set via the `new Route({ prefix: "/1/products" })` I think thats most bullet proof solution to find a "namespace".  

## Remove type
I have removed type, I don't think we need it, its an http based audit trail. 

## Added an objectType 
To capture the what type of the object are we dealing with e.g. is it a user or product or shop ?

## Missing things 
I need endpoints to expose the audit logs for reading
And possible would also be useful to hook it up something like users / auth endpoints ?

kaareal commented on services/api/src/models/audit-log.js:
```diff
@@ -0,0 +1,38 @@
+};
+
+schema.statics.getDiffObject = function getDiffObject(object, fields) {
+  const pathsModified = object.directModifiedPaths();
```
@dominiek / @andrewplummer there is a somewhat of a problem with directModifiedPaths https://mongoosejs.com/docs/api/document.html#document_Document-directModifiedPaths. This works totally fine  ... but naturally doesn't work with auto population (because its not really a build in solution in mongoose). And I don't see an easy solution out of that. 

The best solution for me would be to separate the e.g. user object (auto populate) and the user id.
so we would have 

```
{
  userId: "ObjectId"
  user: {...populated  user object}
}
```

andrewplummer replied:
#1 I'm not sure if the diff object is a requirement (I'm assuming it is to make things simpler from looking up multiple documents)
#2 I wouldn't really expect it to work with autopopulate anyway as an update is by definition not "deep" so I'm wondering if this is really even a problem? You just want to see which IDs changed no?

kaareal replied:
yep only care about objectId, which is what doesn't work, because the auto-populate changes the value of from an objectId to a full populated object. 

still thinking of solutions ... let me know if you have ideas

andrewplummer replied:
there is a `.depopulate` method I know... might be of help?

dominiek replied:
I don't think we need this to work on populated objects.

kaareal replied:
but we do want it to work in the e.g. if the objectId was changed right? (which also not work due to the auto populate replacing the objectId with the full populated object).

dominiek replied:
It would be nice.

dominiek replied:
Not too worried about it now

dominiek commented on services/api/src/models/__tests__/audit-log.js:
```diff
@@ -0,0 +1,98 @@
+      await AuditLog.append('did something', {
+        type: 'security',
+        ctx,
+        objectDiff: { email: user.email },
```
I'm slightly confused by the objectDiff as you have it now.

What we want is:

- For certain API calls, we want to log if certain attributes were changed and what those attributes were changed to
- So for example, if we have a `ClinicalStudy` object with a `numberOfDays` attribute. We would want to specify that we want to track `numberOfDays` and we would want to know the BEFORE and the AFTER state of that attribute.
- This would show up in the audit log as: "Sarah changed Clinical Study X: numberOfDays was changed from 500 to 600"

I'm not sure if we're capturing this here

kaareal replied:
I now have `objectBefore` and `objectAfter` to capture that information.

dominiek commented on services/api/src/routes/auth.js:
```diff
@@ -70,8 +77,23 @@ router
           const authTokenId = generateTokenId();
           await user.updateOne({ loginAttempts: 0, authTokenId });
+
+          await AuditLog.append('authenticated', {
```
`successfully authenticated`

dominiek commented on services/api/src/routes/auth.js:
```diff
@@ -70,8 +77,23 @@ router
+
           ctx.body = { data: { token: createAuthToken(user.id, authTokenId) } };
         } catch (err) {
+          await AuditLog.append('attempted authentication', {
```
`failed authentication`

dominiek reviewed (changes requested):
Almost there!

andrewplummer commented on services/api/src/models/audit-log.js:
```diff
@@ -0,0 +1,75 @@
+  return objectFields;
+};
+
+schema.statics.append = function (activity, ctx, { objectId, objectType, objectBefore, objectAfter, type, user }) {
```
For input this complex I would rather just have a single options object. Making `ctx` a named option would also make it more explicit what it is... so related to what I was saying above maybe could we do something morel like:

```
function({ object, ctx, ...attributes }) {
  if (object) {
    Object.assign(attributes, this.getObjectFields(object));
  }
  if (ctx) {
    Object.assign(attributes, this.getContextFields(ctx);
  }
  return this.create(attributes);
}
```

andrewplummer replied:
Also for naming I feel like I would rather it just be `AuditEntry` (to me the entire "file" is the log, not a single line), and I would kind of rather just do `.create` although I see why you did this...maybe it's difficult or not good to shadow the mongoose method... what do you think?

kaareal replied:
I took out the `ctx` of the options. Because i realized `ctx` its required. 
So the split is simpler now, everything in options `AuditLog.append(activity, ctx, options)` is optional

Re the auditEntry sure we could call it that, but i stayed with auditLog because plural form is auditLogs. But yeah its just a name what ever.

andrewplummer replied:
But `ctx` should not be required no?

kaareal replied:
it should be as it provides the route information

andrewplummer replied:
but you can have an audit trail outside a route... I thought that was part of the requirements

kaareal replied:
```
For each audit trail entry we want to track:
Originating Client IP address
The user ID that was authenticated at the time
The geographical location (bonus for later)
The API call that the mutation happened on
In case of external integrations the API credentials that were used (future)
```

https://bedrockio.slack.com/archives/C013YV1TPLZ/p1618839594326700?thread_ts=1618824211.322200&cid=C013YV1TPLZ

andrewplummer commented on services/api/src/routes/auth.js:
```diff
@@ -33,6 +33,12 @@ router
         ...ctx.request.body,
       });
 
+      await AuditLog.append('registered', ctx, {
```
I'm not sure we want bedrock out of the box to actually use AuditLog as I feel like most projects won't need it... the same goes for the routes... the facility should just be available. Actually it feels more like a plugin but I think we decided against that for some reason.

kaareal replied:
I am fine with removing it, this started out mainly to make it easier to see how to use it. But then I realized this would something i would have enabled by default. E.g. anything around users could be good track for security. 
 
But yeah lets see what @dominiek says

andrewplummer replied:
It takes up a lot of space in the code and is meaningless for most people approaching bedrock out of the box... I would much rather our code show off how simple it is.

andrewplummer commented on services/api/src/routes/products.js:
```diff
@@ -56,7 +57,14 @@ router
+
+    await AuditLog.append('updated product', ctx, {
+      type: 'admin',
+      ...AuditLog.getObjectFields(product, ['shop', 'name']),
```
This is sort of an example of what I was talking about above, we have a static method providing output that goes right back in as input... Unless `getObjectFields` is needed somewhere else it seems better to accept different input and remove some interface in the process.

kaareal replied:
should be addressed now

andrewplummer commented on services/api/src/models/__tests__/audit-log.js:
```diff
@@ -0,0 +1,126 @@
+      expect(log.user.toString()).toBe(user.id);
+    });
+
+    it('dont write if no change happend', async () => {
```
nit: "should not" and "happened"

andrewplummer commented on services/api/src/utils/schema.js:
```diff
@@ -85,6 +85,20 @@ function createSchema(attributes = {}, options = {}) {
   };
 
+  schema.post(/^init|save/, function () {
+    if (!this.$locals.auditOriginal) {
```
so I don't want to beat a dead horse but remind me again what we lose if we don't do this? It just seems like a lot of hacking we're doing here on every save on all models just for the sake of the audit trail... If the entire facility doesn't work without it then maybe I get it but if for example it's just for the sake of making populated objects work slightly better, then I don't think it's worth it.

I'm not really expecting diffs on populated objects anyway I think as populated by definition means they're different models, I'm only interested in what the ref changed to. Remember, as long as we have the refs we could potentially fetch them manually if needed. I feel like this is perfectly fine.

kaareal replied:
All I am doing here is store the original object how it was before it was changed + storing the directModifiedPaths.

Another point I want to clarify, the tracking of populated objects are outside the scope. Nothing is tracking besides the reference (objectId). E.g. Consider the `product.shop` attribute I want to track when the product.shop objectId is changed. 

The reason for this:
First this avoid us from splitting AuditLog.getObjectFields from the AuditLog.append
Before this is what you have write
```
const { user } = ctx.state 
user.assign(ctx.request.body);
const objectFields = AuditLog.getObjectFields(user, ["email"]);
await user.save() 
await AuditLog.append("updated user", ctx, ...objectFields);
```
Now you can write, as the original we still have access to via $locals and we still have access to the what fields has changed.
```
const { user } = ctx.state 
user.assign(ctx.request.body);
await user.save() 
await AuditLog.append("updated user", ctx, ...AuditLog.getObjectFields(user, ["email"]));
```

Think thats worth it for that. I hoped that mongoose would store the original object, but it does not.

The second thing that comes for free is that it also allows me fix handle the objectId issue (its always reported as modified) because i have the original state of the object.

andrewplummer replied:
ok I'm starting to understand so we lose the ability to diff essentially, nothing to do with populating... makes sense I think but a few comments I'll add inline

andrewplummer commented on services/api/src/models/audit-log.js:
```diff
@@ -0,0 +1,75 @@
+        return !object.get(field).equals(get(auditOriginal, field));
+      });
+
+      const after = pick(pick(object.toObject({ depopulate: true }), pathsModified), fields);
```
is this the different from `pick(object, [...pathsModified, ...fields])`?
Also I'm wondering if you ever would actually use `fields` ... the audit trail almost by definition you don't know what you're interested in up front, seems like would always just be interested in all changed fields?

kaareal replied:
`is this the different from pick(object, [...pathsModified, ...fields])?`
Yes. So the logic i am trying todo here is to ensure only get fields that changed. 
But I could possible do an _.intersect with pathsModified and fields to make that a bit cleaner

This is to my understanding one of the requirements, not to track everything but only track specify fields.
cc @dominiek

andrewplummer replied:
well it looks like it's getting fields that changed but also specific ones you can pass in no?

kaareal replied:
Should be addressed now

kaareal commented on services/api/src/models/audit-log.js:
```diff
@@ -0,0 +1,75 @@
+  };
+};
+
+schema.statics.getObjectFields = function getObjectFields(object, fields = [], isNew = false) {
```
I guess it depends => If we have the schema.hooks enabled it could be combined.

andrewplummer replied:
right I see your point around that now... I'm definitely +1 for less interface so the hooks way sounds good to me

andrewplummer commented on services/api/src/utils/schema.js:
```diff
@@ -85,6 +85,20 @@ function createSchema(attributes = {}, options = {}) {
 
+  schema.post(/^init|save/, function () {
+    if (!this.$locals.auditOriginal) {
+      this.$locals.auditOriginal = this.toObject({
```
Can we just call this `original` as this has nothing specific to do with an audit

kaareal replied:
we sure can

andrewplummer commented on services/api/src/utils/schema.js:
```diff
@@ -85,6 +85,20 @@ function createSchema(attributes = {}, options = {}) {
+
+  // keeps the directModifiedPaths around after document.save
+  schema.pre('save', function () {
+    const paths = this.directModifiedPaths() || [];
```
does `directModifiedPaths` ever return undefined?

kaareal replied:
fixed

andrewplummer commented on services/api/src/utils/schema.js:
```diff
@@ -85,6 +85,20 @@ function createSchema(attributes = {}, options = {}) {
+  // keeps the directModifiedPaths around after document.save
+  schema.pre('save', function () {
+    const paths = this.directModifiedPaths() || [];
+    this.$locals.auditPathsModified = uniq([...(this.$locals.pathsModified || []), ...paths]);
```
we probably don't need to store the modified paths themselves just the snapshot on save... something like:

```js
  schema.pre('save', function () {
    const paths = this.directModifiedPaths();
    this.$locals.diff = {
      before: pick(this.$locals.original, paths),
      after: pick(this.toObject({ depopulate: true }), paths),
    };
  });

```

kaareal commented:
I am leaving the AuditEntries hook in the routes to make it more clear how it works. 
I think it make sense to have hooked up for the users / auth endpoints, but happy to remove it.

dominiek commented on services/api/src/roles.json:
```diff
@@ -3,14 +3,16 @@
     "permissions": {
-      "users": "read-write"
+      "users": "read-write",
+      "audit-entries": "read"
```
This should be `auditEntries` - that's the convention we use in all the other projects

dominiek commented:
lgtm
